redis/src/cluster_async/mod.rs

```diff
@@ -0,0 +1,1239 @@
+const DEFAULT_RETRIES: u32 = 16;
+
+/// This is a Redis cluster client.
+pub struct Client {
+    initial_nodes: Vec<ConnectionInfo>,
+    retries: Option<u32>,
+}
+
+impl Client {
+    /// Connect to a redis cluster server and return a cluster client.
+    /// This does not actually open a connection yet but it performs some basic checks on the URL.
+    ///
+    /// # Errors
+    ///
+    /// If it is failed to parse initial_nodes, an error is returned.
+    pub fn open<T: IntoConnectionInfo>(initial_nodes: Vec<T>) -> RedisResult<Client> {
+        let mut nodes = Vec::with_capacity(initial_nodes.len());
+
+        for info in initial_nodes {
+            let info = info.into_connection_info()?;
+            if let ConnectionAddr::Unix(_) = info.addr {
+                return Err(RedisError::from((ErrorKind::InvalidClientConfig,
+                                             "This library cannot use unix socket because Redis's cluster command returns only cluster's IP and port.")));
+            }
+            nodes.push(info);
+        }
+
+        Ok(Client {
+            initial_nodes: nodes,
+            retries: Some(DEFAULT_RETRIES),
+        })
+    }
+
+    /// Set how many times we should retry a query. Set `None` to retry forever.
+    /// Default: 16
+    pub fn set_retries(&mut self, retries: Option<u32>) -> &mut Self {
+        self.retries = retries;
+        self
+    }
+
+    /// Open and get a Redis cluster connection.
+    ///
+    /// # Errors
+    ///
+    /// If it is failed to open connections and to create slots, an error is returned.
+    pub async fn get_connection(&self) -> RedisResult<Connection> {
+        Connection::new(&self.initial_nodes, self.retries).await
+    }
+
+    #[doc(hidden)]
+    pub async fn get_generic_connection<C>(&self) -> RedisResult<Connection<C>>
+    where
+        C: ConnectionLike + Connect + Clone + Send + Sync + Unpin + 'static,
+    {
+        Connection::new(&self.initial_nodes, self.retries).await
+    }
+}
+
+/// This is a connection of Redis cluster.
+#[derive(Clone)]
+pub struct Connection<C = MultiplexedConnection>(mpsc::Sender<Message<C>>);
+
+impl<C> Connection<C>
+where
+    C: ConnectionLike + Connect + Clone + Send + Sync + Unpin + 'static,
+{
+    async fn new(
+        initial_nodes: &[ConnectionInfo],
+        retries: Option<u32>,
+    ) -> RedisResult<Connection<C>> {
+        Pipeline::new(initial_nodes, retries).await.map(|pipeline| {
+            let (tx, mut rx) = mpsc::channel::<Message<_>>(100);
+
+            tokio::spawn(async move {
+                let _ = stream::poll_fn(move |cx| rx.poll_recv(cx))
+                    .map(Ok)
+                    .forward(pipeline)
+                    .await;
+            });
+
+            Connection(tx)
+        })
+    }
+}
+
+type SlotMap = BTreeMap<u16, String>;
+type ConnectionFuture<C> = future::Shared<BoxFuture<'static, C>>;
+type ConnectionMap<C> = HashMap<String, ConnectionFuture<C>>;
+
+struct Pipeline<C> {
```

Comment:
I know this maintains the naming from the `aio` module, but `Pipeline` has two meanings here - both this and `CmdArg::Pipeline` (and `crate::Pipeline`). IMO this duality should be avoided, and the struct renamed to something like `ClusterConnection`, or just `Connection` and rename `Connection` to `NodeConnection`.

Reply:
Agreed that the naming is very confusing. Planning to address the confusion in subsequent PRs.